### ISSUE TITLE
The requested combination of ledgers and assets is not supported.

### ISSUE THREAD

**Issue by yosriady** [comit-network/comit-js-sdk#182]
This is an odd one, but **for some inputs** the cnd HTTP API returns an error when you `sendSwap`:
```
HTTP_URL_CND_1 http://127.0.0.1:51378/swaps/rfc003

{"status":400,"title":"Swap not supported.","detail":"The requested combination of ledgers and assets is not supported."}
```

Here is the swap that triggers the error (this is the request payload passed to the `/swaps/rfc003`endpoint):

```
{
	"alpha_ledger": {
		"name": "ethereum",
		"chain_id": 17
	},
	"beta_ledger": {
		"name": "bitcoin",
		"network": "regtest"
	},
	"alpha_asset": {
		"name": "erc20",
		"token_contract": "0x62d69f6867a0a084c6d313943dc22023bc263691",
		"quantity": "27000000000000000000"
	},
	"beta_asset": {
		"name": "bitcoin",
		"quantity": "376433.99999999994"
	},
	"alpha_ledger_refund_identity": "0xf8960e4BB96108aBD63910A3aC7BCdCd3d53696F",
	"alpha_expiry": 1586409200,
	"beta_expiry": 1586405600,
	"peer": {
		"peer_id": "QmcWiij22RbVB1goi1DR6fz61xRRu1FiXC4Ue6dbiK6E6V",
		"address_hint": "/ip4/127.0.0.1/tcp/9939"
	}
}
```

Am I missing something?

<details>
  <summary>Environment details</summary>

BITCOIN_HD_KEY_0=tprv8ZgxMBicQKsPdP9rgLucRdmHBY7CbM1tEnh3xXrTuf9xZAQabwZUb4zkEkhZ2Vy8YSxZJLLerP3nmDLMA79UiUqVSKQZUDwcZ4MR8AVzF87
BITCOIN_HD_KEY_1=tprv8ZgxMBicQKsPdKmZ5BQWW29cdqLVxDLAzpPtqQ1CuiTSjXBX1KsZ5GJkk1u5vJWuTLT5SgHLnA8LpqG8ohQcSBH7DqSkTQf1KEscPB1UNFU
BITCOIN_P2P_URI=127.0.0.1:51370
ERC20_CONTRACT_ADDRESS=0x62d69f6867a0a084c6d313943dc22023bc263691
ETHEREUM_KEY_0=84f640e67b8062a5e02ef3cc3aff4c8b02e6ea25fd62e16822540bf821e5e9b5
ETHEREUM_KEY_1=79af3b148b215147f627007bbf6d5a0f3ff30ccec5e219343e5c543c50020afe
ETHEREUM_NODE_HTTP_URL=http://127.0.0.1:51356
HTTP_URL_CND_0=http://127.0.0.1:51377
HTTP_URL_CND_1=http://127.0.0.1:51378

</details>

**Comment by yosriady** [comment]
Ok, now it appears to only surface when it's 27 DAI. Not sure what's causing this. 😕

**Comment by yosriady** [comment]
`
cnd:0.4.0
`

**Comment by bonomat** [comment]
@yosriady : I think I found the problem:

```
"beta_asset": {
		"name": "bitcoin",
		"quantity": "376433.99999999994"
	},
```

Same as Ethereum quantity is in `Wei`, Bitcoin quantity is in `Satoshi` and no floating point numbers are allowed here. 

I tried your swap request with: 

```
curl --request POST localhost:18000/swaps/rfc003 -H "Content-Type: application/json" --data '{
        "alpha_ledger": {
                "name": "ethereum",
                "chain_id": 17
        },
        "beta_ledger": {
                "name": "bitcoin",
                "network": "regtest"
        },
        "alpha_asset": {
                "name": "erc20",
                "token_contract": "0x62d69f6867a0a084c6d313943dc22023bc263691",
                "quantity": "27000000000000000000"
        },
        "beta_asset": {
                "name": "bitcoin",
                "quantity": "376433"
        },
        "alpha_ledger_refund_identity": "0xf8960e4BB96108aBD63910A3aC7BCdCd3d53696F",
        "alpha_expiry": 1586409200,
        "beta_expiry": 1586405600,
        "peer": {
                "peer_id": "QmcWiij22RbVB1goi1DR6fz61xRRu1FiXC4Ue6dbiK6E6V",
                "address_hint": "/ip4/127.0.0.1/tcp/9939"
        }
}'

{"id":"4755c25b-00e4-42df-a10a-ab9684079ead"}%
```

You will need to round the quantities and get rid of the floating point.

Please close if this solves your problem.

**Comment by yosriady** [comment]
> @yosriady : I think I found the problem:
> 
> ```
> "beta_asset": {
> 		"name": "bitcoin",
> 		"quantity": "376433.99999999994"
> 	},
> ```
> 
> Same as Ethereum quantity is in `Wei`, Bitcoin quantity is in `Satoshi` and no floating point numbers are allowed here.
> 
> I tried your swap request with:
> 
> ```
> curl --request POST localhost:18000/swaps/rfc003 -H "Content-Type: application/json" --data '{
>         "alpha_ledger": {
>                 "name": "ethereum",
>                 "chain_id": 17
>         },
>         "beta_ledger": {
>                 "name": "bitcoin",
>                 "network": "regtest"
>         },
>         "alpha_asset": {
>                 "name": "erc20",
>                 "token_contract": "0x62d69f6867a0a084c6d313943dc22023bc263691",
>                 "quantity": "27000000000000000000"
>         },
>         "beta_asset": {
>                 "name": "bitcoin",
>                 "quantity": "376433"
>         },
>         "alpha_ledger_refund_identity": "0xf8960e4BB96108aBD63910A3aC7BCdCd3d53696F",
>         "alpha_expiry": 1586409200,
>         "beta_expiry": 1586405600,
>         "peer": {
>                 "peer_id": "QmcWiij22RbVB1goi1DR6fz61xRRu1FiXC4Ue6dbiK6E6V",
>                 "address_hint": "/ip4/127.0.0.1/tcp/9939"
>         }
> }'
> 
> {"id":"4755c25b-00e4-42df-a10a-ab9684079ead"}%
> ```
> 
> You will need to round the quantities and get rid of the floating point.
> 
> Please close if this solves your problem.

Checking.

**Comment by yosriady** [comment]
Yes, I can confirm that it's because there were floats in the swap message. Closing and mitigating on the app side.

**Comment by yosriady** [comment]
Does the cnd endpoint have a payload schema it uses to validate against? If there's like a JSON Schema for the expected payload, we can use it to return more granular error messages to point consumers in the right direction.

**Comment by bonomat** [comment]
Re-opening this and including @thomaseizinger to keep the discussion open.

Do you think this is something we can catch on cnd side?

**Comment by thomaseizinger** [comment]
We don't have a JSON schema for the request body at this stage because it is pretty complicated, for example, the valid values of `alpha_asset.name` depend on the value of `alpha_ledger.name`. We are fixing this by introducing dedicated routes per combination now which will allow us to have a static request body and hence a very easy and precise validation with JSON schema.

**Comment by yosriady** [comment]
Could we can implement client-side validations on the SDK?. For example:

```js
const swap = new Swap(<parameters>);
const isValid = swap.valid?() // Returns true or false, and validation errors.
```

This way, app developers don't have to worry about rolling out their own validation of ledger data, which should be the responsibility of the SDK anyway.
